include/invert_quda.h

```diff
@@ -487,9 +540,16 @@ namespace quda {
 
     virtual void blocksolve(ColorSpinorField &out, ColorSpinorField &in);
 
-    const DiracMatrix &M() { return mat; }
-    const DiracMatrix &Msloppy() { return matSloppy; }
-    const DiracMatrix &Mprecon() { return matPrecon; }
+    virtual void train_param(Solver &, ColorSpinorField &) { errorQuda("NOT implemented."); }
+
+    virtual void solve_and_collect(ColorSpinorField &, ColorSpinorField &, std::vector<ColorSpinorField *> &, int, double) { errorQuda("NOT implemented."); }
```

Comment:
Documentation needed here for what `train_param` and `solve_and_collect` do.  And I'm not sure if I like having these very MADWF concepts in the solver interface.  Wondering if another approach is possible.

Reply:
This is an interface problem that I was having trouble with. Here what we want to do is collect some vectors during the solve, and this interface change that I ended up with, which is ugly and not optimal. But I don't have a better idea.

Reply:
Perhaps this can be a discussion on our next QUDA call

Reply:
Sticking a pin on revisiting this in our next call

Reply:
I don't mind having a `solve_and_collect` specifically---I did something similar with `PrepareSpecialMG` or whatnot with staggered MG. I think we should write it a bit differently, with a `solveAndCollect` and a 'hasSolveAndCollect`, and then we can query for that appropriately elsewhere.

I don't mind the idea of having a `solveAndCollect` in general, who knows what ideas we may have in the future.

Reply:
I might want something like this for DWF MG when I want to shrink `Ls`!